llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td

```diff
@@ -0,0 +1,2359 @@
+    let ReleaseAtCycles = [6, 4, 8];
+}
+
+//===----------------------------------------------------------------------===//
+// Define generic 22 micro-op types
+
+def N3Write_13c_7L_5V_10I : SchedWriteRes<[N3UnitL, N3UnitV, N3UnitI]> {
+    let Latency = 13;
+    let NumMicroOps = 22;
+    let ReleaseAtCycles = [7, 5, 10];
+}
+
+//===----------------------------------------------------------------------===//
+// Define generic 27 micro-op types
+
+def N3Write_11c_6V_9L_12I : SchedWriteRes<[N3UnitV, N3UnitL, N3UnitI]> {
+    let Latency = 11;
+    let NumMicroOps = 27;
+    let ReleaseAtCycles = [6, 9, 12];
+}
+
+// Miscellaneous
+// -----------------------------------------------------------------------------
+
+def : InstRW<[WriteI], (instrs COPY)>;
+
+// Branch Instructions
+// -----------------------------------------------------------------------------
+
+// Branch, immed
+// Compare and branch
+def : SchedAlias<WriteBr, N3Write_1c_1B>;
+
+// Branch, register
+def : SchedAlias<WriteBrReg, N3Write_1c_1B>;
+
+// Branch and link, immed
+// Branch and link, register
+def : InstRW<[N3Write_1c_1B_1S], (instrs BL, BLR)>;
+
+// Arithmetic and Logical Instructions
+// -----------------------------------------------------------------------------
+
+// ALU, basic
+// ALU, basic, flagset
+// Arithmetic, immediate to logical address tag
+// Conditional compare
+// Conditional select
+def : SchedAlias<WriteI, N3Write_1c_1I>;
+
+// ALU, extend and shift
+def : SchedAlias<WriteIEReg, N3Write_2c_1M>;
+
+def N3WriteISReg : SchedWriteVariant<[
+                     SchedVar<IsCheapLSL,  [N3Write_1c_1I]>,
+                     SchedVar<NoSchedPred, [N3Write_2c_1M]>]>;
+
+// Arithmetic, LSL shift, shift <= 4
+// Arithmetic, flagset, LSL shift, shift <= 4
+// Arithmetic, LSR/ASR/ROR shift or LSL shift > 4
+def : SchedAlias<WriteISReg, N3WriteISReg>;
+
+// Convert floating-point condition flags
+def : SchedAlias<WriteSys, N3Write_1c_1I>;
+
+// Flag manipulation instructions
+def : InstRW<[N3Write_1c_1I], (instrs SETF8, SETF16, RMIF, CFINV)>;
+
+// Insert Random Tags
+def : InstRW<[N3Write_2c_1M0], (instrs IRG, IRGstack)>;
+
+// Insert Tag Mask
+// Subtract Pointer
+// Subtract Pointer, flagset
+def : InstRW<[N3Write_1c_1I], (instrs GMI, SUBP, SUBPS)>;
+
+// Logical, shift, no flagset
+def : InstRW<[N3Write_1c_1I],
+             (instregex "^(AND|BIC|EON|EOR|ORN|ORR)[WX]rs$")>;
+
+// Logical, shift, flagset
+def : InstRW<[N3Write_2c_1M], (instregex "^(AND|BIC)S[WX]rs$")>;
+
+// Divide and Multiply Instructions
+// -----------------------------------------------------------------------------
+
+// Integer divides are performed using an iterative algorithm and block any
+// subsequent divide operations until complete.
+
+// Divide, W-form
+def : SchedAlias<WriteID32, N3Write_12c_1M0>;
+
+// Divide, X-form
+def : SchedAlias<WriteID64, N3Write_20c_1M0>;
+
+// Multiply accumulate, W-form
+// Multiply accumulate long
+def : SchedAlias<WriteIM32, N3Write_2c_1M0>;
```

Comment:
A mul (a madd without a addend) seem to be able to execute on either M pipeline.

Reply:
It seems that there are some errors and omissions in "Neoverse N3 Software Optimization Guide".

**From "Neoverse N2 Software Optimization Guide"**:

![{85EEA1F2-D9C3-4BA2-94BE-4F7E0B7B4EDE}](https://github.com/user-attachments/assets/d4d70a73-b98d-4571-8d2c-bd6b4f8686c4)

**From "Neoverse N2 Software Optimization Guide"**:

![{83236E78-EB30-4A76-9FFE-88FAD1ADEF3F}](https://github.com/user-attachments/assets/3285dd7d-0ce9-4a1a-a7f1-e4749b52e804)

Reply:
I need to compare "Neoverse N3 Software Optimization Guide" with "Neoverse N2 Software Optimization Guide" further to fix all similar errors.

Reply:
In the N3 manual it is a bit hidden in Note 3 below the table that says:
```
3. Multiply without accumulate when Ra is ZR (0'b11111), MUL, MNEG, SMULL, SMNEGL, UMULL and UMNEGL
instructions can be executed on utilized pipeline M with an execution throughput of 2.
```

Reply:
Sure.

Reply:
Done.